homeassistant/components/sonos/speaker.py

```diff
@@ -0,0 +1,234 @@
+"""Base class for common speaker tasks."""
+from __future__ import annotations
+
+from asyncio import gather
+import datetime
+import logging
+from typing import Any, Callable
+
+from pysonos.core import SoCo
+from pysonos.events_base import Event as SonosEvent, SubscriptionBase
+from pysonos.exceptions import SoCoException
+
+from homeassistant.core import HomeAssistant, callback
+import homeassistant.helpers.device_registry as dr
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_send,
+    dispatcher_connect,
+    dispatcher_send,
+)
+
+from .const import (
+    DOMAIN as SONOS_DOMAIN,
+    PLATFORMS,
+    SCAN_INTERVAL,
+    SEEN_EXPIRE_TIME,
+    SONOS_CONTENT_UPDATE,
+    SONOS_DISCOVERY_UPDATE,
+    SONOS_ENTITY_CREATED,
+    SONOS_ENTITY_UPDATE,
+    SONOS_GROUP_UPDATE,
+    SONOS_MEDIA_UPDATE,
+    SONOS_PLAYER_RECONNECTED,
+    SONOS_PROPERTIES_UPDATE,
+    SONOS_SEEN,
+    SONOS_STATE_UPDATED,
+    SONOS_VOLUME_UPDATE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SonosSpeaker:
+    """Representation of a Sonos speaker."""
+
+    def __init__(self, hass: HomeAssistant, soco: SoCo):
+        """Initialize a SonosSpeaker."""
+        speaker_info = soco.get_speaker_info(True)
```

Comment:
Pass in `speaker_info` instead of calling it inside of `__init__` (no side effects in init rule)